### ISSUE TITLE
Can't parse output generated inside event handlers.

### ISSUE THREAD

**Issue by brj** [fish-shell/fish-shell#2665]
I realize this is currently not possible, but I needed an issue in order to reference this from other discussions.

For example:

```
function foo -e foo_event
    echo foo event info
end

# ...

emit foo_event # can't parse this!
```

The string `foo event info` will be sent to stdout from another dimension, and there is no way to parse this output.

Using global variables inside the event handlers to communicate information is one possible work around.

**Comment by oranja** [comment]
And then what happens when there are multiple "event handlers"?
I doubt Fish will fire the events asynchronously, but even a synchronous, sequential invocation will give you a mix of outputs that you don't know much about their sources.

Another, IMHO cleaner, approach would be to define a callback event handler and pass the name of the callback with the first event (`emit foo_event <callback>`). When emitting the callback event, the event handler may even specify some kind of identifier to separate it from other registered event handlers. For large outputs it's possible to redirect output from the event handler to a temporary file and pass the path to the file to the callback function/event.

Bottom line, I think Fish already has the tools to get info back from events. Piping to or from an event sounds like an architectural mistake. Emitters and handlers are meant to be loosely coupled by design, as far as I'm aware.

**Comment by bucaran** [comment]
> Bottom line, I think Fish already has the tools to get info back from events. Piping to or from an event sounds like an architectural mistake. 

> Emitters and handlers are meant to be loosely coupled by design, as far as I'm aware.

You are **absolutely** correct.


I implemented the plugin initialization mechanism in OMF/Wahoo using events because (1) it seemed like the right architectural choice and (2) because I thought they were pretty nifty.

Later, while working on Fisherman's predecessor, _Fin_ (which is no more), I implemented a more  complex design that allowed you to plug in your own commands in which there was no need to hardcode a sub-command invocation, essentially a router. 

So, **instead of**:

```fish
function cmd # wa/omf
    # parse options
    switch $cmd
        case a
            cmd_a
        case b
            cmd_b
        case c
            cmd_c
        case d
            cmd_d

    # ... big O

    end
end
```

I had something like:

```fish
# ...
emit sub_command_$cmd $args ...
```

This was a step up in the design, but later in the development I realized that if I could not parse another command's output, this whole "framework" thing was useless.

I removed the event-based design and [implemented](https://github.com/fisherman/fisherman/blob/master/functions/fisher.fish#L65) a similar design using `source` and `eval`.

In case it is not obvious why having commands parse another's output is essential. Let me explain it.

In Fisherman, all commands can read and interpret their standard input one way or another. So, you can plug in the result of a `search` query into `install`, or into itself to further filter the contents. The limit is the sky.

Also, this opens the door to pretty reporters. In Wahoo/OMF, the format / ANSI escapes are hardcoded in the usage help, activity logs, etc. In Fisherman, output is bare by default and it was carefully designed so it was very easy to parse using `cut`, `sed`,  `awk`, etc.

To give only **one** example, all of `fisher` sub-commands, have their auto-complete and help listings generated out of the box by parsing their respective usage texts. If you create a plugin that adds a new command, all you need to do is add a `usage` help like you should if you are writing UNIX tools, and it will be automatically parsed and handled for you.

**Comment by oranja** [comment]
Impressive work and all, but I'm not sure I follow you through and through.

Would you still like to be able to pipe output from events?
Because it seems like what you really want is to be able to call a function/command indirectly. So you could simply use `cmd_$cmd --usage`, without using `eval`.

Do correct me if I'm wrong and clarify.

**Comment by bucaran** [comment]
> Would you still like to be able to pipe output from events?

No, I would not like that.

Like you already noted:

> I doubt Fish will fire the events asynchronously, but even a synchronous, sequential invocation will give you a mix of outputs that you don't know much about their sources.

Exactly. Parsing the output generated inside an event handler, does not make a lot of sense.

> Because it seems like what you really want is to be able to call a function/command indirectly. 

Yes.


> So you could simply use cmd_$cmd --usage, without using eval.

No. You can't.

<img width="730" alt="screen shot 2016-01-17 at 6 31 36 pm" src="https://cloud.githubusercontent.com/assets/8317250/12376786/912bd72e-bd48-11e5-8d5d-40188f0cdcd3.png">

**Comment by oranja** [comment]
>> So you could simply use cmd_$cmd --usage, without using eval.

> You can't.

I know, I know. I only meant that this is something you would like to be able to do. (Edited my previous post to make it clearer).

I stumbled across the same problem too and wished it would not have to include `eval` (which I have been taught to associate with 'evil') , but it's not an accident that Fish doesn't allow you to use a variable in a command/function name:

> Since the use of variables as commands in regular shells is allowed, it is impossible to reliably check the syntax of a script. [...] fish solves this by disallowing variables as commands. Anything you can do with variables as commands can be done in a much cleaner way using either the eval command or by using functions.

(From http://lwn.net/Articles/136232/, referenced in #154)

Related: #154, #2427

**Comment by bucaran** [comment]
> which I have been taught to associate with 'evil'

Depends? If the language had a way to store functions in variables and call them like in say JavaScript, this issue would not exist.

Very nice article <http://lwn.net/Articles/136232/>! I felt I went back in time

**Comment by floam** [comment]
Should this be left open?

**Comment by brj** [comment]
No, let's close it.
